design/BufferOperations.md

```diff
@@ -0,0 +1,150 @@
+`getPointer` returns an ArrayBuffer representing the buffer data if the object is in the available state, null otherwise.
+
+`WebGPUMappedMemory` is also `then`-able, meaning that it acts like a Javascript `Promise`:
+
+```
+partial interface WebGPUMappedMemory {
+    Promise then(WebGPUMappedMemorySuccessCallback success,
+                 optional WebGPUMappedMemoryErrorCallback error);
+};
+```
+
+This acts like a `Promise<ArrayBuffer>.then` that is resolved on the Javascript task boundary in which the implementation detects the GPU is done with the buffer.
+On that boundary:
+
+ - The `WebGPUMappedMemory` goes in the available state.
+ - If the `WebGPUMappedMemory` was created via `WebGPUBuffer.mapWrite`, its content is cleared to 0.
+ - `success` is called with the content of the memory as an argument.
+
+If `success` hasn't been called when the WebGPUMappedMemory gets invalidated (meaning the object is still in the pending state), `error` is called instead. When `WebGPUMappedMemory` goes from the available state to the invalidated state, the `ArrayBuffer` for its content gets neutered. The return value of `then` acts like the return value of `Promise.then`.
+
+The `ArrayBuffer` of a `WebGPUMappedMemory` created from a `mapWrite` is where the application should write the data and its content is made available to the buffer when the `WebGPUMappedMemory` is invalidated (i.e. `WebGPUBuffer.unmap` is called).
+
+## Immediate data upload
+
+Buffer mapping is the path with the least number of copies but it is often useful to upload data to a buffer *right now*, if only for debugging.
+A `WebGPUBuffer` operation is provided that takes an ArrayBuffer and copies its content at an offset in the buffer.
+
+```
+partial interface WebGPUBuffer {
+    void setSubData(ArrayBuffer data, u32 offset);
+}
+```
+
+This operation acts as if it was done after all previous "device-level" commands and before all subsequent "device-level" commands.
+"Device level" commands are all commands not buffered in a `WebGPUCommandBuffer`, and include `WebGPUQueue.submit`.
+The content of `data` is only read during the call and can be modified by the application afterwards.
+The following must be true or a validation error occurs:
+
+ - The buffer must have been created with the `WebGPUBufferUsage.TRANSFER_DST` usage flag.
+ - `offset + data.length` must not overflow and be at most the size of the buffer.
+ - Depending on the design of memory barriers, the buffer must be, or allowed to be in the `WebGPUBufferUsage.TRANSFER_DST` usage.
+   - In particular the buffer must not be currently mapped.
+
+## Unused designs
+
+### Persistently mapped buffer
+
+Persistently mapped buffer are when the result of mapping the buffer can be kept by the application while the buffer is in use by the GPU.
+We didn't find a way to have persistently mapped buffers and at the same time keep things data race free between the CPU and GPU.
```

Comment:
If this section will be committed, you should state that "we" is "The WebGPU Community Group"